### PR TITLE
Decrease Pipe lock held time to reduce contention

### PR DESCRIPTION
**Pull request by benaadams**
50% of `AdvanceReader` is `ResetMemory()` and there is also lock contention

image

Schedule the next step prior to returning the memory so its ready to run earlier and reset the memory outside the lock (so its held for less time); then return to the pipe pool under lock again.

/cc @davidfowl @halter73 @pakrym @sebastienros

### CODE REVIEW

**Comment by pakrym** [comment]
Would be nice to see some perf numbers.
Re-taking lock twice in the single method is not great I also wonder if it is worth polling more than 1 buffer segment at all.

**Comment by benaadams** [comment]
>  I also wonder if it is worth polling more than 1 buffer segment at all.

Changed to single segment which means lock isn't needed to be retaken

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 1]
```diff
@@ -463,17 +457,22 @@ private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Buf
...
+            // Reset the segments outside lock
+            BufferSegment returnLast = null;
+            while (returnStart != null && returnStart != returnEnd)
+            {
+                returnLast = returnStart;
+                returnStart = returnStart.NextSegment;
+                returnLast.ResetMemory();
+            }
+
+            ReturnSegment(returnLast);
```
Comment (pakrym) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Would returning the segment so late after `TrySchedule` that it won't be able to ever pick it up in time would make polling effectively useless?

Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Also moving to 1 segment causes questions around `DefaultMinimumSegmentSize` which is based on number of pooled segments; so will skip it in this PR

**Comment by ahsonkhan** [comment]
@dotnet-bot test NETFX x86 Release Build

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 1 above)
Comment (benaadams) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Probably can still do something a bit better here though

**Comment by benaadams** [comment]
Hmm... might have over thunk this...

Going back to the original of just returning outside lock

**Comment by benaadams** [comment]
Not convinced by the double lock, but there are some pref improvements that will open in another PR

**Comment by benaadams** [comment]
More strightforward perf tweaks https://github.com/dotnet/corefx/pull/33917

**Pull request closed without merging**